# LAB ASSIGNMENT

**Name:** SHUBHADIP CHAKRABARTI

**Registration Number:** 19BAI1113

**Objective:** Application of Convolution Neural Netwrosk on an Image Dataset

**Dataset Used**: Retinal Disease Dataset (Image Dataset)

**Dataset Procurement Source:** The dataset used in this notebook is procured from Kaggle Website 

We now mount the Google Drive to enable us to import the image dataset

### Mounting Google Drive into the Notebook

In [1]:
#Mounting the Google Drive into the Notebook
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Having successfully mounted the Google Drive into the python Notebook, we now import necessary modules and functions, for carrying out the exercise, into the notebook

### Importing Necessary Modules and Functions

In [2]:
#import OS
import os

#import numpy
import numpy as np

#import pandas
import pandas as pd

#import glob
import glob

#import Keras
from keras.preprocessing.image import ImageDataGenerator,load_img, img_to_array
from keras.models import Sequential, Model
from keras.layers import Conv2D, MaxPooling2D
from keras.layers import Activation, Dropout, Flatten, Dense, AvgPool2D, MaxPool2D
from keras.models import Sequential, Model

#import Tensorflow
import tensorflow as tf

#import Adam
from tensorflow.keras.optimizers import Adam

Having imported all the necessary functions into the Notebook, we now move on to the execution of the exercise.

### Importing Dataset into the Notebook

In [3]:
#dataset path
dataset = "/content/drive/MyDrive/Computer Vision in Healthcare Application/Lab/Dataset"

### Defining Parameters

In [4]:
#shape of image
image_width = 256
image_height = 256
channels = 3

input_shape = (image_width, image_height, channels)

#number of output classes
classes = 39

#batch size
batch_size = 64

Having defined the necessary parameters, let us now define the Neural Network

### Neural Network

We aim to design a Sequential Neural Network. 

In [5]:
#Neural Network
model = Sequential()

model.add(Conv2D(32, (3, 3), input_shape=input_shape))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(64, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(64,(3,3)))
model.add(Activation("relu"))

model.add(Conv2D(250,(3,3)))
model.add(Activation("relu"))
  
model.add(Conv2D(128,(3,3)))
model.add(Activation("relu"))
model.add(AvgPool2D(2,2))

model.add(Conv2D(64,(3,3)))
model.add(Activation("relu"))
model.add(AvgPool2D(2,2))

model.add(Conv2D(256,(2,2)))
model.add(Activation("relu"))
model.add(MaxPool2D(2,2))
    
model.add(Flatten())        #flattern
model.add(Dense(32))        #dense layer
model.add(Dropout(0.25))    #dropout layer
model.add(Dense(1))
model.add(Activation("sigmoid"))    

Having defined the Model, we now compile the model.

### Compile the Model

In [6]:
#Compile the Model
model.compile(loss='categorical_crossentropy',    #loss function defined
              optimizer='rmsprop',    #optimizer
              metrics=['accuracy'])   #metrics for evaluation

In [7]:
#Summarize the Model
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 254, 254, 32)      896       
                                                                 
 activation (Activation)     (None, 254, 254, 32)      0         
                                                                 
 max_pooling2d (MaxPooling2D  (None, 127, 127, 32)     0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 125, 125, 64)      18496     
                                                                 
 activation_1 (Activation)   (None, 125, 125, 64)      0         
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 62, 62, 64)       0         
 2D)                                                    

### Split of Dataset

In [8]:
train_datagen = ImageDataGenerator(rescale=1./255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    validation_split=0.3)

train_generator = train_datagen.flow_from_directory(
    dataset,
    target_size=(image_height, image_width),
    batch_size= batch_size,
    subset='training')

validation_generator = train_datagen.flow_from_directory(
    dataset, 
    target_size=(image_height, image_width),
    batch_size= batch_size,
    shuffle= False,
    subset='validation')

Found 716 images belonging to 39 classes.
Found 284 images belonging to 39 classes.


We now carry out Modelling.

### Modelling

We here train the above defined architecture of the Convolution Neural Network (CNN) on the imported Retinal Disease Image Dataset.

In [10]:
#Train the Model
history = model.fit_generator(
    train_generator,
    steps_per_epoch = train_generator.samples // batch_size,
    validation_data = validation_generator, 
    validation_steps = validation_generator.samples // batch_size,
    epochs = 10)    #run the model for 10 epochs

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  


Epoch 1/10
11/11 [==============================] - 315s 28s/step - loss: 3.7086 - accuracy: 0.7367 - val_loss: 3.6455 - val_accuracy: 0.9745
Epoch 2/10
11/11 [==============================] - 291s 26s/step - loss: 3.5746 - accuracy: 0.6338 - val_loss: 3.4684 - val_accuracy: 0.9744
Epoch 3/10
11/11 [==============================] - 292s 29s/step - loss: 3.5245 - accuracy: 0.7384 - val_loss: 3.4506 - val_accuracy: 0.9744
Epoch 4/10
11/11 [==============================] - 292s 26s/step - loss: 3.5480 - accuracy: 0.7509 - val_loss: 3.4439 - val_accuracy: 0.9744
Epoch 5/10
11/11 [==============================] - 292s 26s/step - loss: 3.5449 - accuracy: 0.8093 - val_loss: 3.4414 - val_accuracy: 0.9744
Epoch 6/10
11/11 [==============================] - 290s 26s/step - loss: 3.4630 - accuracy: 0.8745 - val_loss: 3.4040 - val_accuracy: 0.9744
Epoch 7/10
11/11 [==============================] - 292s 26s/step - loss: 3.4788 - accuracy: 0.9077 - val_loss: 3.4357 - val_accuracy: 0.9744
Epoch 

We have successfully trained the CNN Model for the Retinal Image Dataset. We observe that we have acheived an accuracy of 87.85% and validation accuracy of 97.44%. It is also noted that we have achieved a loss of 3.47 and validation loss of 3.36 from the trained CNN Model.

### Performance of the Trained Model

In [11]:
print("Training_accuracy", history.history['accuracy'][-1])
print("Validation_accuracy", history.history['val_accuracy'][-1])

Training_accuracy 0.8784804344177246
Validation_accuracy 0.9743589758872986


We observe that we have achieved an accuracy of 87.85% and validation accuracy of 97.44%.

## Conclusion

We observe that we have successfully trained the Convolution Neural Network on a Retinal Disease Image Dataset.

As observed, we have successfully obtained a 

        1. Training Accuracy of 87.84%
        2. Validation Accuracy of 97.43%